File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Murmur332.java
package com.freight.hash;

/** murmur3_32 over raw bytes. */
public final class Murmur332 implements HashAlgorithm {

    @Override
    public String name() {
        return "murmur3_32";
    }

    @Override
    public long apply(byte[] data) {
        int state = 0x5F3A1C7D;
        int blocks = data.length / 4;
        for (int i = 0; i < blocks; i++) {
            int k = (data[i * 4] & 0xFF)
                    | ((data[i * 4 + 1] & 0xFF) << 8)
                    | ((data[i * 4 + 2] & 0xFF) << 16)
                    | ((data[i * 4 + 3] & 0xFF) << 24);
            k *= 0xCC9E2D51;
            k = Integer.rotateLeft(k, 15);
            k *= 0x1B873593;
            state ^= k;
            state = Integer.rotateLeft(state, 13);
            state = state * 5 + 0xE6546B64;
        }
        int tail = 0;
        int remainder = data.length & 3;
        if (remainder >= 3) {
            tail ^= (data[blocks * 4 + 2] & 0xFF) << 16;
        }
        if (remainder >= 2) {
            tail ^= (data[blocks * 4 + 1] & 0xFF) << 8;
        }
        if (remainder >= 1) {
            tail ^= (data[blocks * 4] & 0xFF);
            tail *= 0xCC9E2D51;
            tail = Integer.rotateLeft(tail, 15);
            tail *= 0x1B873593;
            state ^= tail;
        }
        state ^= data.length;
        state ^= state >>> 16;
        state *= 0x85EBCA6B;
        state ^= state >>> 13;
        state *= 0xC2B2AE35;
        state ^= state >>> 16;
        return state & 0xFFFFFFFFL;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/hashx/murmur3_32.go
package hashx

// Murmur332 computes the murmur3_32 checksum over raw bytes.
func Murmur332(data []byte) uint64 {
	state := uint32(0x5F3A1C7D)
	blocks := len(data) / 4
	for i := 0; i < blocks; i++ {
		k := uint32(data[i*4]) | uint32(data[i*4+1])<<8 | uint32(data[i*4+2])<<16 |
			uint32(data[i*4+3])<<24
		k *= 0xCC9E2D51
		k = (k << 15) | (k >> 17)
		k *= 0x1B873593
		state ^= k
		state = (state << 13) | (state >> 19)
		state = state*5 + 0xE6546B64
	}
	tail := uint32(0)
	remainder := len(data) & 3
	if remainder >= 3 {
		tail ^= uint32(data[blocks*4+2]) << 16
	}
	if remainder >= 2 {
		tail ^= uint32(data[blocks*4+1]) << 8
	}
	if remainder >= 1 {
		tail ^= uint32(data[blocks*4])
		tail *= 0xCC9E2D51
		tail = (tail << 15) | (tail >> 17)
		tail *= 0x1B873593
		state ^= tail
	}
	state ^= uint32(len(data))
	state ^= state >> 16
	state *= 0x85EBCA6B
	state ^= state >> 13
	state *= 0xC2B2AE35
	state ^= state >> 16
	return uint64(state)
}
